source/common/upstream/upstream_impl.h

```diff
@@ -470,12 +466,23 @@ class ClusterImplBase : public Cluster, protected Logger::Loggable<Logger::Id::u
   virtual void startPreInit() PURE;
 
   /**
-   * Called by every concrete cluster when pre-init is complete. At this point, shared init takes
-   * over and determines if there is an initial health check pass needed, etc.
+   * Called by every concrete cluster when pre-init is complete. At this point,
+   * shared init starts sds_init_manager_ initialization and determines if there
+   * is an initial health check pass needed, etc. onPreInitComplete() may be called
+   * multiple times, but sds_init_manager_ only can be called once.
```

Comment:
In what cases is onPreInitComplete() called multiple times? That doesn't make sense but I don't recall the details here.

Reply:
StrictDnsClusterImpl::ResolveTarget::startResolve() is invoked by timer resolve_timer_, which calls parent_.onPreInitComplete(). There is already a ClusterImplBase::initialization_started_ which protects initialization be called multiple times. I need to move the checking of  initialization_started_ from onInitDone() to onPreInitCOmplete().